File: 543-Diameter-of-Binary-Tree/py/main.py
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn diameter_of_binary_tree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut res: Vec<i32> = vec![0];   

        pub fn dfs(root: Option<Rc<RefCell<TreeNode>>>, res: &mut Vec<i32>) -> i32 {
            match root {
            None => -1,
            Some(n) => {
                let left = dfs(n.borrow().left.clone(), res);
                let right = dfs(n.borrow().right.clone(), res);

                let other = 2 + left + right;
                res[0] = res[0].max(other);

                1 + left.max(right)
            }
            }
        }         
        dfs(root, &mut res);
        res[0]
    }
}

    

File: 543-Diameter-of-Binary-Tree/go/main.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// Function to calculate the diameter of a binary tree
func diameterOfBinaryTree(root *TreeNode) int {
    var res = [1]int{0}

    var dfs func(node *TreeNode) int
    dfs = func(node *TreeNode) int {
        if node == nil {
            return -1
        }

        leftHeight := dfs(node.Left)
        rightHeight := dfs(node.Right)

        res[0] = max(res[0], 2+leftHeight+rightHeight)

        return 1 + max(leftHeight, rightHeight)
    }

    dfs(root)
    return res[0]
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}
